\documentclass{bioinfo}
\copyrightyear{2012}
\pubyear{2012}

\usepackage{subfig}
\usepackage{listings}
\usepackage{ifthen}

\lstset{language=Java,
morendkeywords={String, Throwable}
captionpos=b,
basicstyle=\scriptsize\ttfamily,%\bfseries
stringstyle=\color{darkred}\scriptsize\ttfamily,
keywordstyle=\color{royalblue}\bfseries\ttfamily,
ndkeywordstyle=\color{forrestgreen},
numbers=left,
numberstyle=\scriptsize,
% backgroundcolor=\color{lightgray},
breaklines=true,
tabsize=2,
frame=single,
breakatwhitespace=true,
identifierstyle=\color{black},
% morecomment=[l][\color{forrestgreen}]{//},
% morecomment=[s][\color{lightblue}]{/**}{*/},
% morecomment=[s][\color{forrestgreen}]{/*}{*/},
commentstyle=\ttfamily\itshape\color{forrestgreen}
% framexleftmargin=5mm,
% rulesepcolor=\color{lightgray}
% frameround=ttff
}

% MACROS:

\newcommand{\TODO}[1]{\textcolor{red}{\textbf{#1}}}

\hyphenation{
% TODO: insert hypens here!
}

% some nice colors
\definecolor{royalblue}{cmyk}{.93, .79, 0, 0}
\definecolor{lightblue}{cmyk}{.10, .017, 0, 0}
\definecolor{forrestgreen}{cmyk}{.76, 0, .76, .45}
\definecolor{darkred}{rgb}{.7,0,0}
\definecolor{winered}{cmyk}{0,1,0.331,0.502}
\definecolor{lightgray}{gray}{0.97}

\begin{document}
\firstpage{1}

\title[SBML Simulation Core Library]{SBML Simulation Core Library: the Java Library for numerical computation with SBML}
\author[Dr\"ager \textit{et~al.}]{Andreas Dr\"ager\,$^{1,*}$,
Roland Keller\,$^{1}$, 
Alexander D\"orr\,$^{1}$,
Akito Tabira\,$^{2}$,
Akira Funahashi\,$^{2}$,
\TODO{Nicolas Rodriguez\,$^{3}$,
Nicolas Le Nov\`{e}re\,$^{3}$} and
Andreas Zell\,$^1$\footnote{to whom correspondence should be addressed}}
\address{$^{1}$Center for Bioinformatics Tuebingen (ZBIT), University of
Tuebingen, T\"ubingen, Germany\\
$^{2}$Keio University, Graduate School of Science and Technology, Yokohama,
Japan\\
$^{3}$European Bioinformatics Institute, Wellcome Trust Genome Campus,
Hinxton, Cambridge, UK\\}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}

\section{Motivation:}
Java\texttrademark{} programming language
\TODO{Text Text Text Text Text Text  Text Text Text Text Text Text}

\section{Results:}
\TODO{Text Text Text Text Text Text  Text Text Text Text Text Text}

\section{Availability:}
Source code, binaries, and documentation for SBML Simulation Core can
be freely obtained under the terms of the LGPL 3.1 from the website
\href{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator}{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator}.

\section{Contact:}
% TODO: We need a mailing list for this!
\href{mailto:andreas.draeger@uni-tuebingen.de}{andreas.draeger@uni-tuebingen.de}

\section{Supplementary information:}
% TODO: Provide additional material
Supplementary data are available at Bioinformatics online.

\end{abstract}

\section{Introduction}

The modeling language SBML (Systems Biology Markup Language,
\citealt{Hucka2003}) constitutes an important de facto standard for the
exchange of biochemical network models.
SBML defines a set of data structures and provides rules about how to interpret
and simulate these kinds of models.

A model stored in SBML may combine an ordinary differential equation system,
which is the basis for numerical simulation, with additional elements such as
rules and events.  These elements further influence the system. For instance,
an event takes place if a certain trigger condition becomes true. Whenever this
happens, event assignments may change the values of model components, such as
parameter values or compartment sizes. Rules can directly assign new values to
their objectives, e.g., the concentration of a reacting species.

In many cases, model parameters with uncertain values, such as kinetic
constants, need to be estimated using heuristic optimization procedures in
order to minimize the distance between model output and experimental
measurement data.

For model building, simulation, and parameter calibration processes, dedicated
software implementations are indispensable. We here present a simulator that
interprets the content of such a model and predicts the dynamic behavior of the
modelÕs components. It is based on the Javaª library JSBML \citep{Draeger2011b},
a specifically developed data structure to read and write models from and into
SBML files and to deal with their structure in memory.

\TODO{The program SBMLsimulator consists of two parts: Firstly, a generic solver
core library, which is completely decoupled from any graphical user interface and
can hence easily be integrated as an API (Application Programming Interface)
into third-party programs. The core can be downloaded at}
\TODO{Insert Uni Tuebingen website here!}
\href{http://sourceforge.net/projects/sbml-simulator/}{http://sourceforge.net/projects/sbml-simulator/}
under the terms of the LGPL Version 3. 
\TODO{Describe this as a potential application.}
Secondly, a graphical and command-line user interface that provides
a connection to the heuristic optimization framework EvA2 \citep{Kron10EvA2}. The combination
of SBMLsimulator and EvA2 \citep{Kron10EvA2} estimates the values of all parameters with
respect to given time-series of metabolite or gene expression values. The
simulation core will also be integral part of the widely used program
CellDesigner version 4.2 \citep{Funahashi2003}.


\section{Approach}

\TODO{Text Text Text Text Text Text  Text Text Text Text Text Text}

\begin{methods}
\section{Methods}

\TODO{Text Text Text Text Text Text  Text Text Text Text Text Text}

\begin{itemize}
\item for bulleted list, use itemize
\end{itemize}

\TODO{Text Text Text Text Text Text  Text Text Text Text Text Text}

%\begin{table}[!t]
%\processtable{This is table caption\label{Tab:01}}
%{\begin{tabular}{llll}\toprule
%head1 & head2 & head3 & head4\\\midrule
%row1 & row1 & row1 & row1\\
%row4 & row4 & row4 & row4\\\botrule
%\end{tabular}}{This is a footnote}
%\end{table}

\end{methods}

%\begin{figure}[!tpb]%figure1
%%\centerline{\includegraphics{fig01.eps}}
%\caption{Caption, caption.}\label{fig:01}
%\end{figure}

\section{Discussion}

\TODO{Text Text Text Text Text Text  Text Text Text Text Text Text}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%     please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}}
%     as it may ignore the figures.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section{Conclusion}

\TODO{Text Text Text Text Text Text  Text Text Text Text Text Text}


\section*{Acknowledgement}

The authors are grateful to Beky Kotcon, Samantha Mesuro, Daniel Rozenfeld, Anak
Yodpinyanee, Andres Perez, Eric Doi, Richard Mehlinger, Steven Ehrlich, Martin
Hunt, George Tucker, Peter Scherpelz, Aaron Becker, Eric Harley, and Chris
Moore, Harvey Mudd College, Claremont, California, USA, for providing an
Java implementation of Rosenbrock's numerical ODE solver. 

\paragraph{Funding\textcolon} 
The Federal Ministry of Education and Research (BMBF, Germany) in the project
Virtual Liver Network (grant number 0315756).

\paragraph{Conflict of Interest\textcolon} none declared.

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{document}

\end{document}
