\documentclass{bioinfo}
\copyrightyear{2012}
\pubyear{2012}

\usepackage{subfig}
\usepackage{listings}
\usepackage{ifthen}
\usepackage[english]{babel}
\usepackage[normalem]{ulem}

\lstset{language=Java,
morendkeywords={String, Throwable}
captionpos=b,
basicstyle=\scriptsize\ttfamily,%\bfseries
stringstyle=\color{darkred}\scriptsize\ttfamily,
keywordstyle=\color{royalblue}\bfseries\ttfamily,
ndkeywordstyle=\color{forrestgreen},
numbers=left,
numberstyle=\scriptsize,
% backgroundcolor=\color{lightgray},
breaklines=true,
tabsize=2,
frame=single,
breakatwhitespace=true,
identifierstyle=\color{black},
% morecomment=[l][\color{forrestgreen}]{//},
% morecomment=[s][\color{lightblue}]{/**}{*/},
% morecomment=[s][\color{forrestgreen}]{/*}{*/},
commentstyle=\ttfamily\itshape\color{forrestgreen}
% framexleftmargin=5mm,
% rulesepcolor=\color{lightgray}
% frameround=ttff
}

% MACROS:

\newcommand{\TODO}[1]{\textcolor{red}{\textbf{#1}}}
\newcommand{\AbstractDESSolver}{\texttt{Abstract\-DES\-Solver}}
\newcommand{\OverdeterminationValidator}{\texttt{Overdetermination\-Validator}}
\newcommand{\SBMLinterpreter}{\texttt{SBML\-interpreter}}
\newcommand{\FirstOrderSolver}{\texttt{First\-Order\-Solver}}
\newcommand{\AbstractIntegrator}{\texttt{AbstractIntegrator}}
\newcommand{\MultiTable}{\texttt{Multi\-Table}}
\newcommand{\Block}{\texttt{Block}}

\hyphenation{
  % TODO hypens for regular words
  im-ple-men-ta-tions
  gra-phi-cal
  bench-mark-ed
}

% some nice colors
\definecolor{royalblue}{cmyk}{.93, .79, 0, 0}
\definecolor{lightblue}{cmyk}{.10, .017, 0, 0}
\definecolor{forrestgreen}{cmyk}{.76, 0, .76, .45}
\definecolor{darkred}{rgb}{.7,0,0}
\definecolor{winered}{cmyk}{0,1,0.331,0.502}
\definecolor{lightgray}{gray}{0.97}

\begin{document}
\firstpage{1}

\title[Simulation Core Library]{Simulation Core Library: the Java
library for numerical computation in systems biology} \author[Dr\"ager
\textit{et~al.}]{%
Andreas Dr\"ager\,$^{1,}$\footnote{to whom correspondence should be
addressed}\hspace{.3em},
Roland Keller\,$^{1,}$\dag,
Alexander D\"orr\,$^{1,}$\footnote{Authors with equal
contribution}\hspace{.3em}, Akito Tabira\,$^{2}$,
Akira Funahashi\,$^{2}$,
Michael J. Ziller\,$^{3}$,
Nicolas Rodriguez\,$^{4}$,
Nicolas Le Nov\`{e}re\,$^{4}$
Andreas Zell\,$^{1,*}$}
\address{$^{1}$Center for Bioinformatics Tuebingen (ZBIT), University of
Tuebingen, T\"ubingen, Germany\\
$^{2}$Keio University, Graduate School of Science and Technology, Yokohama,
Japan\\
$^{3}$Department of Stem Cell and Regenerative Biology, Harvard University,
Cambridge, MA, USA\\
$^{4}$European Bioinformatics Institute, Wellcome Trust Genome Campus,
Hinxton, Cambridge, UK}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}
\section{Motivation:}
Dynamic simulation of biological phenomena are key aspects of
research in systems biology. However, it is often difficult to use available implementations of numerical methods as a backend for custom-made programs.
\section{Results:}
The Simulation Core Library is a community-driven project that provides a large
collection of numerical solvers and a sophisticated interface hierarchy for the
definition of custom differential equation systems. It is entirely
implemented in Java\texttrademark{} without the necessity to
include any platform-dependent wrappers or libraries, and does not depend on
any commercial library.
%, and can be used on every operating system for which a JVM
%is available.
It already includes an efficient and exhaustive implementation of methods to
interpret the content of models encoded in SBML using the JSBML project.
To demonstrate its capabilities, it has been benchmarked against the entire SBML
Test Suite and %also been used to simulate 
all models of BioModels Database.
\section{Availability:}
Source code, binaries, and documentation can be freely obtained under the terms
of the LGPL 3.1 from the website
\href{https://sourceforge.net/projects/simulation-core-library}{https://sourceforge.net/projects/simulation-core-library}.

\section{Contact:}
\href{mailto:andreas.draeger@uni-tuebingen.de}{andreas.draeger@uni-tuebingen.de}

%\section{Supplementary information:}
% TODO: Provide additional material
%Supplementary data is available at Bioinformatics online.

\end{abstract}

\section{Introduction}

As part of the movement towards predictable biology, modeling, 
simulation, and computer analysis of biological networks, have become integral
parts of modern biological research. XML-based standard description formats
such as the Systems Biology Markup Language (SBML, \citealt{Hucka2003}) or
CellML \citep{Lloyd2004} provide means to encode biological network models, and interpret them in terms of a differential equation system, with
additional structures such as discrete events and algebraic equations.
Software libraries for reading and manipulating the content of
these formats are available.
However, a prerequisite for model analysis, simulation, and calibration (e.g., the
estimation of parameter values), is a multiple-purpose and 
efficient numerical solver library that has been designed with the
requirements of biological network models in mind.

Many stand-alone programs providing these important features come with graphical user interfaces.
\marginpar{I would avoid naming any software here. Such a list can only be incomplete and is not necessary}
\sout{, for instance, the Virtual Cell \citep{Loew2001}, iBioSim \citep{Myers2009},
PottersWheel \citep{Maiwald2008}, COPASI \citep{Hoops2006}, SBToolbox2
\citep{SBT_Schmidt2006}}
%, or the Systems Biology Workbench with Roadrunner (SBW,
%\citealt{Bergmann06}). 
However, the vast majority of the internal solvers for
these systems is part of a larger software suite and can therefore not
be easily integrated into custom programs. Some are implemented in programing
languages which are either platform-dependent (e.g., C or C++) and/or require
a commercial license (e.g., MATLAB\texttrademark{}) for their execution.
%
%The modeling language SBML (Systems Biology Markup Language,
%\citealt{Hucka2003}) constitutes an important \emph{de facto} standard for the
%exchange of biochemical network models.
%SBML defines a set of data structures and provides rules about how to interpret
%and simulate these kinds of models.
%
%Models in systems biology may combine an ordinary differential equation system,
%which is the basis for numerical simulation, with additional elements such as
%rules and events.  These elements further influence the system. 
%For instance,
%an event takes place if a certain trigger condition becomes true. Whenever this
%happens, event assignments may change the values of model components, such as
%parameter values or compartment sizes. Rules can directly assign new values to
%their objectives, e.g., the concentration of a reacting species.
%

\sout{The SBML ODE Solver Library \citep{Machne2006}, which is written in C,
%and based on the libSBML library \citep{Bornstein2008}, 
provides such a simulation routine based on the SUNDIALS differential equation
solver.}
\marginpar{I would avoid mentioning it here but keep it for the discussion. It makes the current work look like a reimplementation in Java}

In contrast, we present here the platform-independent Simulation Core
Library. This generic library is completely decoupled from any graphical user
interface and can therefore easily be integrated into third-party programs. 
It contains several Ordinary Differential Equation (ODE)
solvers and an interpreter for SBML models. It is the first simulation library
 based on the Java library JSBML \citep{Draeger2011b}, specifically
developed for reading and writing models from and into SBML
files and to deal with their structure in memory.

%Secondly, a graphical and command-line user interface that provides
%a connection to the heuristic optimization framework EvA2 \citep{Kron10EvA2}.
% The combination of SBMLsimulator and EvA2 \citep{Kron10EvA2} estimates the values of all parameters with
%respect to given time-series of metabolite or gene expression values. 

\begin{methods}
\section{Implementation}
\begin{figure}
%\centerline{
%  \subfloat[Solvers.]{
%    \label{fig:Solvers}
%    \includegraphics[width=.5\textwidth]{img/Solvers.pdf}
%  }
%  \hfill
%  \subfloat[Differential equation systems.]{
%    \label{fig:DESystems}
%    \includegraphics[width=.5\textwidth]{img/SBMLInterpreter.pdf}
%  }
%}
\centering{\includegraphics[width=.5\textwidth]{img/UML_Solvers_and_Systems_1.pdf}}
\caption[Architecture of the Simulation Core Library]{Architecture of
the Simulation Core Library (slightly simplified). Numerical methods are
strictly separated from differential equation systems. The upper part displays
the unified type hierarchy of all currently included numberical integration
methods. The middle part shows the interfaces defining several
special types of the differential equations to be solved by the numerical
methods.
The class \SBMLinterpreter{} (bottom part) implements all of these interfaces
with respect to the information content of a given SBML model. Similarly, an
implementation of further data formats can be included into the library.}
\label{fig:Architecture}
\end{figure}
Fig.~\ref{fig:Architecture} displays the architecture of the library. All the
solver classes are derived from the abstract class \AbstractDESSolver.
Several solvers of the Apache Commons Math library (version 3.0) are integrated
with the help of wrapper classes. Numerical methods and the actual differential
equation systems are strictly separated. The class \MultiTable{} stores the
results of a simulation within its \Block{} data structures. 

The abstract description of differential equation systems with the help of
several distinct interfaces makes possible to decouple it from a particular type
of biological networks. It is therfore possible to pass an instance of an
interpreter for a particular model description format to any
available solver.\marginpar{I would not quote SBML and CellML. CellML is actually not supported at the moment}

A specialized interpreter class is required for the evaluation of a biological model. This interpretation is the most time consuming step of the integration procedure.
This is why efficient and clearly organized data structures are required to
ensure a high performance of the overall library. The interpretation of SBML
models is split between evaluation of events and rules, computation of stoichiometric information, and computation of the current
values for all model components (such as species and compartments).

For a given state of the ODE system\marginpar(the state includes all variables including time), the class \SBMLinterpreter, responsible for the evaluation of models encoded in SBML returns the current set of
derivatives of the variables. It is connected to an efficient
interpreter for MathML expressions that are contained in kinetic laws, rules
and events. The nodes of the syntax tree of those expressions depend on the
current simulation time and the given values of the ODE system. If the time or
any of these values has changed, the value of the node has to be recalculated.
At the beginning of the simulation the syntax trees of all kinetic laws, rules
and events are restructured and merged to one large tree that contains
equivalent nodes only once. This leads to a decreasing computation time during
the simulation.

An extremely important aspect in the interpretation of SBML models is the
determination of the exact time at which an event occurs, as this can have a
high influence on the precision of the values of the system variables. We have
therefore adapted the Rosenbrock solver \citep{Kotcon2011}, which is an
integrator with an adaptive step size, to a very precise timing of the events.
Rosenbrock's method is well-suited even for stiff systems.

Algebraic rules are transferred to assignment rules before the simulation: One
of the variables contained in an algebraic rule is chosen as the variable of the
created assignment rule and the equation is solved by this variable. The
\OverdeterminationValidator{} in JSBML helps to ensure that no conflicts occur
between any of the chosen variables.

The simulation algorithm now proceeds as follows: In each time step the ODE
solver gets the current set of values of the variables as its initial values and
calculates the values for the next time point. After that events
and rules are processed, that can change the values. The calculated values are
then the initial values for the next time step. The event processing of the
Rosenbrock solver is different from that of the other solvers, as it
is directly integrated in the solver class and influences the adaptation of its
step size. This makes the processing of the events extremely precise.
\end{methods}

%\begin{table}[!t]
%\processtable{This is table caption\label{Tab:01}}
%{\begin{tabular}{llll}\toprule
%head1 & head2 & head3 & head4\\\midrule
%row1 & row1 & row1 & row1\\
%row4 & row4 & row4 & row4\\\botrule
%\end{tabular}}{This is a footnote}
%\end{table}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%     please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}}
%     as it may ignore the figures.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section{Results and conclusion}
The SBML implementation has been successfully passed the
SBML Test Suite using the Rosenbrock solver.
%(see
%\href{http://sbml.org/Software/SBML_Test_Suite}{http://sbml.org/Software/SBML\_Test\_Suite}):
Furthermore, it solved nearly all models from the
\href{http://biomodels.net}{BioModels.net} database (406 of 409 models from
release 21, \citealp{Novere2006a}).
Therefore, the Simulation Core Library is an efficient Java tool for the
simulation of differential equation systems given in systems biology. It can be
easily integrated in larger applications.
The widespread program CellDesigner version~4.2 \citep{Funahashi2003} uses it
already as one of its simulation libraries.
The stand-alone application SBMLsimulator (available at
\href{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator}{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator})
provides a convenient graphical user interface for the simulation of SBML
models and uses this library as its computational backend.
The abstract program structure of the Simulation Core Library supports the
integration of additional model formats besides its SBML implementation, for
instance CellML. To this end, it is only necessary to implement a suitable
interpreter class.

\section*{Acknowledgement}

The authors are grateful to Beky Kotcon, Samantha Mesuro, Daniel Rozenfeld, Anak
Yodpinyanee, Andres Perez, Eric Doi, Richard Mehlinger, Steven Ehrlich, Martin
Hunt, George Tucker, Peter Scherpelz, Aaron Becker, Eric Harley, and Chris
Moore, Harvey Mudd College, USA, for providing a Java implementation of
Rosenbrock's method, and to Michael T. Cooling, University of Auckland, New
Zealand.

\paragraph{Funding\textcolon} 
The Federal Ministry of Education and Research (BMBF, Germany) in the project
Virtual Liver Network (grant number 0315756).

\paragraph{Conflict of Interest\textcolon} none declared.

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{document}

\end{document}
