%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to 
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <14 August 2007>            %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%				      	               %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
\usepackage{pdflscape}
\usepackage{booktabs}
\usepackage{color}
\usepackage[table]{xcolor}

\definecolor{royalblue}{cmyk}{.93, .79, 0, 0}
\definecolor{lightblue}{cmyk}{.10, .017, 0, 0}
\definecolor{darkgreen}{rgb}{0,.7,0}
\definecolor{darkred}{rgb}{.7,0,0}
\definecolor{lightgray}{gray}{0.97}

%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     


\def\includegraphic{}
\def\includegraphics{}



\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 

%Review style settings
%\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

%New style setting
\newenvironment{bmcformat}{\baselineskip20pt\sloppy\setboolean{publ}{false}}{\baselineskip20pt\sloppy}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%  Customized tags and styles
%
\usepackage{hyperref}
\usepackage{listings}
\usepackage[english]{babel}
%\usepackage[normalem]{ulem}
\usepackage{xcolor}

\hyphenation{
  % TODO hypens for regular words
  im-ple-men-ta-tions
  gra-phi-cal
  bench-mark-ed
}

% some nice colors
\definecolor{royalblue}{cmyk}{.93, .79, 0, 0}
\definecolor{lightblue}{cmyk}{.10, .017, 0, 0}
\definecolor{forrestgreen}{cmyk}{.76, 0, .76, .45}
\definecolor{darkred}{rgb}{.7,0,0}
\definecolor{winered}{cmyk}{0,1,0.331,0.502}
\definecolor{lightgray}{gray}{0.97}

\newcommand{\TODO}[1]{\textcolor{red}{\textbf{#1}}}
\newcommand{\AbstractDESSolver}{\texttt{Abstract\-DES\-Solver}}
\newcommand{\AlgebraicRule}{\texttt{Algebraic\-Rule}}
\newcommand{\AssignmentRule}{\texttt{Assignment\-Rule}}
\newcommand{\OverdeterminationValidator}{\texttt{Overdetermination\-Validator}}
\newcommand{\SBMLinterpreter}{\texttt{SBML\-interpreter}}
\newcommand{\FirstOrderSolver}{\texttt{First\-Order\-Solver}}
\newcommand{\AbstractIntegrator}{\texttt{AbstractIntegrator}}
\newcommand{\MultiTable}{\texttt{Multi\-Table}}
\newcommand{\Block}{\texttt{Block}}
\newcommand{\jlibsedml}{\texttt{jlibsedml}}
\newcommand{\EventInProgress}{\texttt{Event\-In\-Progress}}
\newcommand{\ASTNodeInterpreter}{\texttt{ASTNode\-In\-terpreter}}
\newcommand{\SBMLEventInProgressWithDelay}{\texttt{SBML\-Event\-In\-Progress\-With\-Delay}}

% the derivative symbol for differential equations
\newcommand{\D}{\mathrm{d}}


\lstset{language=Java,
morendkeywords={String, Throwable}
captionpos=b,
basicstyle=\scriptsize\ttfamily,%\bfseries
stringstyle=\color{darkred}\scriptsize\ttfamily,
keywordstyle=\color{royalblue}\bfseries\ttfamily,
ndkeywordstyle=\color{forrestgreen},
numbers=left,
numberstyle=\scriptsize,
% backgroundcolor=\color{lightgray},
breaklines=true,
tabsize=2,
frame=single,
breakatwhitespace=true,
identifierstyle=\color{black},
% morecomment=[l][\color{forrestgreen}]{//},
% morecomment=[s][\color{lightblue}]{/**}{*/},
% morecomment=[s][\color{forrestgreen}]{/*}{*/},
commentstyle=\ttfamily\itshape\color{forrestgreen}
% framexleftmargin=5mm,
% rulesepcolor=\color{lightgray}
% frameround=ttff
}
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% Begin ...
\begin{document}
\begin{bmcformat}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{The systems biology Simulation Core Algorithm and its implementation}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Ensure \and is entered between all but   %%
%% the last two authors. This will be       %%
%% replaced by a comma in the final article %%
%%                                          %%
%% Ensure there are no trailing spaces at   %% 
%% the ends of the lines                    %%     	
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\author{%
Roland Keller$^{1}$, %\email{Roland Keller - roland.keller@uni-tuebingen.de}, 
Alexander D\"orr$^{1}$, %\email{Alexander D\"orr - alexander.doerr@uni-tuebingen.de},  
Akito Tabira$^{2}$, %\email{Akito Tabira - mythosil_full_throttle@i.softbank.jp}
Akira Funahashi$^{2}$, %\email{Akira Funahashi - funa@bio.keio.ac.jp}
Michael J. Ziller$^{3}$, %\email{Michael J. Ziller - michael_ziller@harvard.edu}
Richard Adams$^{4}$, %\email{Richard Adams - richard.adams@ed.ac.uk}
Nicolas Rodriguez$^{5}$, %\email{Nicolas Rodriguez - rodrigue@ebi.ac.uk}
Nicolas Le Nov\`{e}re$^{5}$, %\email{Nicolas Le Nov\`{e}re - nicolas.lenovere@babraham.ac.uk}
Hannes Planatscher$^{6}$, %\email{Hannes Planatscher - Hannes.Planatscher@nmi.de}
Andreas Zell$^{1}$, %\email{Andreas Zell - andreas.zell@uni-tuebingen.de}
and Andreas Dr\"ager$^{1}$\correspondingauthor\email{Andreas Dr\"ager -
andreas.draeger@uni-tuebingen.de}%
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address{%
\iid(1)Center for Bioinformatics Tuebingen (ZBIT), University of
Tuebingen, T\"ubingen, Germany
\iid(2)Keio University, Graduate School of
Science and Technology, Yokohama, Japan 
\iid(3)Department of Stem Cell and Regenerative Biology, Harvard University,
Cambridge, MA, USA
\iid(4)SynthSys Edinburgh, CH Waddington Building, University of Edinburgh,
Edinburgh EH9 3JD, UK
\iid(5)European Bioinformatics Institute, Wellcome Trust Genome Campus, Hinxton,
Cambridge, UK
\iid(6)Natural and Medical Sciences Institute at the University of Tuebingen,
Reutlingen, Germany}

\maketitle

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%  
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%   
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\begin{abstract}
        % Do not use inserted blank lines (ie \\) until main body of text.
Dynamic simulation of models describing biological phenomena is a key aspect of
research in systems biology. 
This article describes an efficient and exhaustive algorithm to interpret and
solve the differential equation systems of models encoded in Systems Biology
Markup Language (SBML) and its implementation as part of the Simulation Core
Library for numerical computation in systems biology.

This library is a community-driven project that provides a large collection of
numerical solvers and a sophisticated interface hierarchy
for the definition of custom differential equation systems. It is hence possible
to extend this library with additional interpreters for other systems biological
standard formats.
The Simulation Core Library is intended to be used as a backend for custom-made
programs.
It is entirely implemented in Java\texttrademark, using the JSBML project,
without the necessity to include any platform-dependent wrappers or libraries,
and does not depend on any commercial licenses.
%, and can be used on every operating system for which a JVM
%is available.
To demonstrate its capabilities, it has been tested with the
entire SBML Test Suite and %also been used to simulate 
all models of BioModels Database.

Source code, binaries, and documentation can be freely obtained under the terms
of the LGPL version~3 from the website
\href{http://sourceforge.net/projects/simulation-core/}{http://sourceforge.net/projects/simulation-core/}.
Feature requests, bug reports, or any further discussion can be directed to the mailing list
\href{mailto:simulation-core-development@lists.sourceforge.net}{simulation-core-development@lists.sourceforge.net}.
%
%\section{Supplementary information:}
% TODO: Provide additional material
%Supplementary data is available at Bioinformatics online.     
\end{abstract}



\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %% 
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%




%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background}

\TODO{Is more explanation why this is needed and comparison to other existing
approaches necessary?}

As part of the movement towards quantitative biology, modeling, 
simulation, and computer analysis of biological networks have become integral
parts of modern biological research. Ambitious national and international
research projects such as the Virtual Liver Network \cite{Holzhuetter2012}
strive to derive even organ-wide models of biological systems that include all
kinds of processes taking place at several levels of detail. Large-scale efforts
like this require a strong collaboration between various research groups,
including experimenters and modelers. The exchange, storage, interoperability
and the possibility to combine models have been recognized as key aspects of
this endeavor \cite{Liebermeister2009sta}.

XML-based standard description formats \cite{Bray2000} such as the Systems
Biology Markup Language (SBML, \cite{Hucka2004}) or CellML \cite{Lloyd2004}
enable encoding of quantitative biological network models.
To facilitate sharing and reuse of the models, online data bases such as the
BioModels database \cite{Novere2006a} or the CellML model repository
\cite{Lloyd2008} provide large collections of published models in various
formats.
Software libraries for reading and manipulating the content
of these formats are also available \cite{Bornstein2008, Miller2010,
Draeger2011b} and a large variety of programs supports these model formats.

\TODO{Should we describe and cite genome-scale models and their size. Can our solver also
work with large-scale models such as HepatoNet1?}

The models encoded in these formats can be interpreted in terms of differential
equation systems, with additional structures such as discrete events and
algebraic equations. 
For efficient model analysis, simulation, and calibration (e.g.,
the estimation of parameter values) multiple-purpose and efficient numerical
solver library that has been designed with the requirements of biological
network models in mind is prerequisite.
Althought the language specifications of SBML \cite{Hucka2001, Hucka2003,
Finney2003a, Finney2006, Hucka2007, Hucka2008, Hucka2010a} 
and CellML \cite{Cuellar2006} describe the semantics of models in these formats
and their interpretation, the algorithmic implementation is still not
straightforward.

Both communities offer standardized and manually derived benchmark test in order
to evaluate the quality of simulation results, because it has been recognized that in many cases
different solver implementations lead to divergent results \cite{Bergmann2008}.

In this work we address the question of how to precisely solve models encoded in
the SBML format, supporting all levels and versions. To this end, we here
describe a precise solver algorithm. As a reference implementation, we
introduce an exhaustive implementation in the Java\texttrademark{} programming
language. The algorithm described in this paper is, however, not limited to any
particular programing language. It is also important to note that the
interpretation of these models must be strictly separated from the numerical
method that solves the implicated differential equation system. In this way, a
similar approach can also be done for CellML or other formats.

%
%The modeling language SBML (Systems Biology Markup Language,
%\cite{Hucka2003}) constitutes an important \emph{de facto} standard for the
%exchange of biochemical network models.
%SBML defines a set of data structures and provides rules about how to interpret
%and simulate these kinds of models.
%
%Models in systems biology may combine an ordinary differential equation system,
%which is the basis for numerical simulation, with additional elements such as
%rules and events.  These elements further influence the system. 
%For instance,
%an event takes place if a certain trigger condition becomes true. Whenever this
%happens, event assignments may change the values of model components, such as
%parameter values or compartment sizes. Rules can directly assign new values to
%their objectives, e.g., the concentration of a reacting species.
%
%
The Simulation Core Library presented here is a platform-independent,
well-tested generic library that is completely decoupled from any graphical
user interface and can therefore easily be integrated into third-party programs.
It comprises several Ordinary Differential Equation (ODE)
solvers and an interpreter for SBML models. It is the first simulation library
based on JSBML \cite{Draeger2011b}. 
% the Java library JSBML 
%
%Secondly, a graphical and command-line user interface that provides
%a connection to the heuristic optimization framework EvA2 \cite{Kron10EvA2}.
% The combination of SBMLsimulator and EvA2 \cite{Kron10EvA2} estimates the values of all parameters with
%respect to given time-series of metabolite or gene expression values. 
%
Furthermore, the Simulation Core Library contains classes to both export
simulation configurations to SED-ML (Simulation Experiment Description Markup Language,
\cite{Waltemath2011}), and facilitate the re-use and reproduction of these
experiments by executing SED-ML files.

\section*{Results and discussion}

The mathematical structure of a reaction network comprises a stoichiometric
matrix $\mathbf{N}$, whose rows correspond to the reacting species $\vec{S}$
within the system, whereas its columns represent the reactions, i.e., bio-transformations,
in which these species participate.
The velocities $\vec{\nu}$ of the reactions $\vec{R}$ determine the rate of
change of the species' amounts:
\begin{equation}
\frac{\D}{\D t}\vec{S} = \mathbf{N}\vec{\nu}(\vec{S}, t, \mathbf{W}, \vec{p})\,.
\label{eq:SpeciesChange}
\end{equation}
The parameter vector $\vec{p}$ contains rate constants and other, often
constant, quantities that influence the reaction velocities.
According to \cite{Liebermeister2006, Liebermeister2010} the modulation matrix
$\mathbf{W}$ is defined as a matrix of size $|\vec{R}|\times|\vec{S}|$
containing the type of the regulatory influences of the species on
the reactions, e.g., competetive inhibition or physical stimulation.
Integrating Equation~\ref{eq:SpeciesChange} yields the predicted amounts of the
species at each time point of interest within the interval $[t_0, t_T]$:
\begin{equation}
\vec{S} = \int_{t_0}^{t_T} \mathbf{N}\vec{\nu}(\vec{S}, t, \mathbf{W}, \vec{p})
\D t\,.
\end{equation}
Depending on the units of the species, the same notation can also express the
change of the species' concentrations.
In this simple case, solving the homogeneous ordinary
differential equation system~\ref{eq:SpeciesChange} can be done in a
straightforward way using many (numerical) differential equation solvers.
The non-linear form of the kinetic equations in the vector function $\vec{\nu}$
constitutes the major difficulty for this endeavor and is often the reason why
an analytical solution of these systems is not possible or very hard to achieve.
Generally, differential equation systems describing biological networks are,
however, inhomogenious systems with a higher complexity.
Solving systems encoded in SBML can be seen as computing the solution of the following
equation:
\begin{equation}
\vec{Q} = \int_{t_0}^{t_T} \mathbf{N}\vec{\nu}(\vec{Q}, t, \mathbf{W},
\vec{p})\D t + \int_{t_0}^{t_T} \vec{g}(\vec{Q}, t)\D t + \vec{e}(\vec{Q}, t) +
\vec{r}(\vec{Q}, t)\,.
\label{eq:QuantityValue}
\end{equation}
The vector $\vec{Q}$ of quantities contains the sizes of the
compartments $\vec{C}$, amounts (or concentrations) of reacting species
$\vec{S}$, and the values of all global model parameters $\vec{P}$.
It should be noted that these models may contain local parameters $\vec{p}$ that
influence the reaction velocities, but which are not part of the global parameter
vector $\vec{P}$, and hence also not part of $\vec{Q}$.
In the general case of Equation~\ref{eq:QuantityValue}, not all species' amounts
can be computed by integrating the transformation $\mathbf{N}\vec{\nu}$: The
change of some model quantities may be given in form of rate rules (function
$\vec{g}(\vec{Q}, t)$) that must be integrated separately.
Species, whose amounts are determined by rate rules, must not participate in any
reaction and do hence not have a corresponding entry in the stoichiometric
matrix $\mathbf{N}$.
Thereby, a rate rule directly gives the rate of change of some quantity:
\begin{equation}
\frac{\D}{\D t}\vec{Q} = \vec{g}(\vec{Q}, t)\,.
\end{equation}
In addition, SBML introduces the concept of events $\vec{e}(\vec{Q}, t)$ and
$\vec{r}(\vec{Q}, t)$ assignment rules.
An event can directly manipulate the value of a quantity, for instance,
set the size of a certain compartment to half of its current size, as soon as a
trigger condition becomes satisfied.
An assignment rule also influences the absolute value of some model.
A further concept of SBML models are algebraic rules, which are equations that
must evaluate to zero at all times during the simulation of the model.
These rules can be solved to determine the values of quantities, whose values
are not determined by other constructs.
In this way, conservation relations or other complex interrelations can be
expressed in a very convenient way.
With the help of a bipartite matching and a subsequent conversion it is possible
to turn algebraic rules into assignment rules and hence include these into the
term $\vec{r}(\vec{Q}, t)$.
Such a transformation, however, requires symbolic computation and is
hence a complicated endeavor.
In case that the system under study operates at multiple time scales, i.e., it
contains a fast and a slow sub-system, a separation of the system is necessary
leading to differential algebraic equations.
Some species can be declared to operate at the system's boundaries, assuming a
constant pool of their amounts or concentrations.
Care must als be taken with respect to the units of the species, because under
certain condition division or multiplication with the sizes of their surrounding
compartments is necessary in order to ensure the consistent interpretation of
the models. 
For all these reasons, solving Equation~\ref{eq:QuantityValue} is much more 
complicated than computing the solution of the simple Equation~\ref{eq:SpeciesChange}
alone.

From the perspective of software engineering, a strict separation of the 
interpretation of the model and the numerical treatment of the differential
equation system is necessary to ensure that regular numerical methods can be used
to solve Equation~\ref{eq:QuantityValue}.
In order to efficiently compute this solution, multiple pre-processing steps are
required, such as the conversion of algebraic rules into assignment rules, or
avoiding repeated re-computation of intermediate results.
The next sections will give a detailed explanation of the necessary steps to
solve these systems and how to perform their numerical integration with standard
numerical solvers.


\subsection*{Initialization}

At the beginning of the simulation the values of species, parameters and
compartments are set to the initial values as given in the model.
All kinetic laws of the reactions, assignment rules, transformed algebraic
rules (see below), initial assignments, event assignments and rate rules
are integrated into a directed acyclic syntax graph.
This graph is hence the result of merging of the abstract syntax trees of
all those elements.
Equivalent elements are only contained once.
This significantly decreases the computation time needed for the
evaluation of these syntax trees during the simulation.
An example for such a syntax graph is given in Figure~1.

After the creating of the abstract syntax graph, the initial assignments, and the
assignment rules (including transformed algebraic rules) are processed.
Deviating initial values defined by initial assignments and are now assigned.


\subsection*{Solving algebraic rules}
In order to deal with algebraic rules in an SBML model, these have to be converted to assignment rules.
In every equation of an algebraic rule, there should be at least one variable, 
whose value is not yet defined through other equations in model. This variable has to be determined 
for the purpose of interpreting the regarding algebraic rule. 
At first, a bipartite graph according to the SBML specifications is generated. This graph is used to 
compute a matching, using the algorithm from Hopcroft and Karp
\cite{hopcroft1973n}. The initial greedy matching is extended with the use of augmenting paths. This process is
repeated until no more augmenting paths can be found. Per definition, this results in a maximal matching. 
As stated in the SBML specifications, if any equation vertex remains unconnected after augmenting the 
matching as far as possible, the model is considered overdetermined and thus is not a valid SBML model. 
But if we are indeed looking at a valid SBML model, every assignment rule is converted into an algebraic rule. 
To this end, the mathematical expression is transformed into an equation with the target variable on its left-hand side.
The left-hand side is represented by the respective variable vertex, to which the considered algebraic rule has been matched.
Figure~2 displays the described algorithm in form of a flow chart.


\subsection*{Event handling}
An event in SBML is a function that is executed depending on whether a trigger condition switches from false to
true.
In addition, a delay may postpone the actual execution of the event assignments to a later point in time.
With the release of SBML Level~3 Version~1, the processing of events has been raised to a higher level of complexity.
Before 3.1 it was sufficient to determine, when an events triggers and when its assignmens are to be exceuted.
In 3.1 just a few new language elements have been added but with an huge impact on how to handle events:
The order in which events have been processed used to be at programmers discretion, but now it is given by the event's priority.
Coordinating the sequence, in which events are to be executed, has now become the crucial part of event handling. 
Furthermore, there exists the option to cancel an event during the time since its trigger 
has been activated and the actual time when the scheduler picks the event for execution.

For every time step, the events to be executed are a union of two subsets of the set of all events.
On the one hand, there are the events, whose trigger has been activated at the given time and which are to be evaluated without delay.
On the other hand, there are events that triggered at some time point before, and whose delay reaches till the current point in time.
For every element of the resulting set of events their priority rule must be evaluated.
An event with the highest priority is randomly chosen for execution.
All other events could be handled in the same manner.
However, the assignment of the first event can change the priority or even the trigger condition of the events that have not yet been executed.
Therefore, the trigger of non persistent events and the priority of the remaining events have to be evaluated again.
In this case, another event that has now the highest priority is chosen.
This process repeats itself until no further events to be executed are left.
Figure~3 shows the slightly simplified algorithm for event processing at a specific point in time.

\subsection*{Calculation of the derivatives for a certain point in time}
For given values $\vec{Q}$ at a point in time $t$ the currend vector of derivatives is calculated as follows:
First all derivatives in the vector $\dot{\vec{Q}}$ are set to zero.
Then the rate rules are processed, which can change $\dot{\vec{Q}}$.
After that for each reaction the velocity, which depends on $\vec{Q}$ at time $t$, is computed with the help of the syntax graph.
Then the derivatives of all species that participate in the current reaction need to be updated (see the flow chart in Figure~4).

\subsection*{Integrated calculation for a certain time step including event processing}
The Rosenbrock solver is a solver that can adapt its step size if events are happening (see Figure~5).
For a certain time begin and time end and a vector $\vec{Q}$ it determines a new vector at a point in time that is determined by the current adaptive step size.
It refers to the already described calculation of the derivatives for a certain point in time in this step.

After that the events and the assignment rules are processed at the new point in time.
If this last step causes a change in $\vec{Q}$, the adaptive step size is decreased and
the calculation is repeated until the minimum step size is reached or the processing of events and rules does not change $\vec{Q}$ any more.
Hence, the time, at which an event takes place, is precisely determined.

\TODO{Discuss implementation in Java, i.e., explain the result, the Simulation Core Library: Where to download and some other details, performance etc.}
%Why no heap is used for events 

\subsection*{Application to published models}

To illustrate the capabilities of the Simulation Core Library, two example
networks from the BioModels database have been selected for simulation:
Biomodel 206 by Wolf \emph{et al.} \cite{Wolf2000} and
Biomodel 390 by Arnold and Nikoloski \cite{Arnold2011}.

The model by Wolf \emph{et al.} \cite{Wolf2000} mimics glycolytic oscillations.

\TODO{Describe the precise settings how the simulation was performed: Which
solver, how many steps, error tolerance etc.}

By comparing a large collection of previous models of the Calvin-Benson
cycle, Arnold and Nikoloski created a quantitative consensus model that
comprises eleven species, six reactions, and one assignment rule
\cite{Arnold2011}.
All kinetic equations within this model call specialized function definitions.
\TODO{Figure~XXX} shows the simulation results for this model.

\subsection*{Comparison to existing solvers}

We selected all SBML-capable simulation programs, libraries, and frameworks from
\url{http://sbml.org/SBML_Software_Guide/SBML_Software_Matrix} (October
8\textsuperscript{th} 2012), which satisfy the following criteria:
\begin{itemize}
  \item The last updated version was released after the final release of
  the specification for SBML Level 3 Version 1 Core, i.e., October
  6\textsuperscript{th} 2010.
  \item Open-source software
  \item No dependency on commercial products that are not freely available
  (e.g., MATLAB\texttrademark{} or Mathematica\texttrademark)
  \item Support for SBML Level 3. 
\end{itemize}
Table~1a and Table~1b summarize the comparison of six of these programs. Only three applications pass the four exemplary models (966, 988, 1083, 1000) of the SBML Test Suite (version 2.0.2): iBioSim \cite{Myers2009}, BioUML \cite{Kolpakov2011}, \cite{Kolpakov2006}, and LibSBMLSim \TODO{Citation}. iBioSim does not yet support the simulation of SBML models containing the csymbol delay feature. LibSBMLSim, which is a simulation API written in C, can only read models given in SBML Level 2 Version 4 and SBML Level 3. So BioUML is the only simulation tool that passes all models of the SBML Test Suite for all levels and versions of SBML.

Many stand-alone programs providing simulation of SBML come with
graphical user interfaces.
For instance, the Virtual Cell \cite{Loew2001}, JSim \cite{Beard2012a}, iBioSim \cite{Myers2009},
PottersWheel \cite{Maiwald2008}, COPASI \cite{Hoops2006}, SYCAMORE \cite{Weidemann2008}, SBToolbox2
\cite{SBT_Schmidt2006}, JWS Online \cite{Olivier2004},or the Systems Biology Workbench with Roadrunner (SBW, \cite{Bergmann06}). 
However, the vast majority of the internal solvers for these systems are part of
larger software suites and can therefore not be easily integrated into custom
programs. Some are implemented in programming languages that are either
platform-dependent (e.g., C or C++) and/or require a commercial license (e.g.,
MATLAB\texttrademark{}) for their execution.
The SBML ODE Solver Library \cite{Machne2006}, which is written in C,
and based on the libSBML library \cite{Bornstein2008}, 
provides such a simulation routine based on the SUNDIALS differential equation
solver.

\section*{Conclusions}
The SBML implementation has successfully passed the
SBML Test Suite (version 2.0.2) using Rosenbrock solver. The results are shown in Table~2. So BioUML and the Simulation Core Library are the only SBML applications that pass the whole Test Suite for all SBML levels and versions. Compared to BioUML the Simulation Core Library simulates all models of the Test Suite much faster.
%(see
%\href{http://sbml.org/Software/SBML_Test_Suite}{http://sbml.org/Software/SBML\_Test\_Suite}):
Furthermore, Simulation Core Library solved 99.06\,\% of the models from the
\href{http://biomodels.net}{BioModels.net} database (release 23,
\cite{Novere2006a}) simulating from $t=0$ to $t=10$ using Rosenbrock solver and a step size of $0.1$.
Therefore, the Simulation Core Library is an efficient Java tool for the
simulation of differential equation systems used in systems biology. It can be
easily integrated in larger applications. For instance,
CellDesigner version~4.2 \cite{Funahashi2003} already uses it as one of its simulation libraries.
The stand-alone application SBMLsimulator (available at
\href{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator}{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator})
provides a convenient graphical user interface for the simulation of SBML
models and uses it as a computational backend.
The abstract class structure of the library supports the integration of
additional model formats, such as CellML, besides its SBML implementation. To
this end, it is only necessary to implement a suitable interpreter class.

%The SBML ODE Solver Library \cite{Machne2006}, which is written in C,
%and based on the libSBML library \cite{Bornstein2008}, 
%provides such a simulation routine based on the SUNDIALS differential equation
%solver.

By including support for the emerging standard SED-ML, we hope to facilitate the
exchange, archival and reproduction of simulation experiments performed using
the Simulation Core Library.

\section*{Methods}

\subsection*{Implementation}

All the solver classes are derived from the abstract class \AbstractDESSolver{}
(Figure~6).
Several solvers of the Apache Commons Math library (version 3.0) are integrated
with the help of wrapper classes. Numerical methods and the actual differential
equation systems are strictly separated. The class \MultiTable{} stores the
results of a simulation within its \Block{} data structures. 
%
The abstract description of differential equation systems, with the help of
several distinct interfaces, makes possible to decouple them from a particular
type of biological network. It is therefore possible to pass an instance of an
interpreter for a respective model description format to any available solver.
%\marginpar{I would not quote SBML and CellML. CellML is
% actually not supported at the moment}
%
%A specialized interpreter class is required for the evaluation of a biological
%model. 
This interpretation is the most time consuming step of the integration procedure.
This is why efficient and clearly organized data structures are required to
ensure a high performance of the overall library. The interpretation of SBML
models is split between evaluation of events and rules, computation of
stoichiometric information, and computation of the current values for all model
components (such as species and compartments).
%
For a given state of the ODE system, the class \SBMLinterpreter{}, responsible
for the evaluation of models encoded in SBML returns the current set of
time-derivatives of the variables.
It is connected to an efficient MathML interpreter of the expressions contained
in kinetic laws, rules and events (\ASTNodeInterpreter{}). The nodes of the syntax graph for those
expressions depend on the current state of the ODE system. If the state has
changed, the values of the nodes have to be recalculated (see Results).

%
An important aspect in the interpretation of SBML models is the
determination of the exact time at which an event occurs, as this influences
the precision of the system's variables. We therefore adjusted the Rosenbrock
solver \cite{Kotcon2011}, an integrator with an adaptive step size, to a very
precise timing of the events.
%\sout{Rosenbrock's method is well-suited even for stiff systems.}
%
In addition to events, Rules are also treated during integration. Basically, they are Events that occur at every given point in time and are therefore processed in the same manner. For every object of the type \AlgebraicRule{}, a new \AssignmentRule{} object is generated by means of the preceding bipartite matching. They represent only temporary rules, that are incorporated in the simulation process but do not influence the model in the SBML file.
%\marginpar{This is only valid for polynomes. And for those,
%assignmentRules should have been used anyway. How do-you proceed for
%cos(x)=0?: Yes, that's true. People should use assignments there, but the
%AlgebraicRules in the test case are all of the type described here. I am
%currently not sure if we could solve cos(x) = 0 or similar cases.}
%

In the \SBMLinterpreter{} events are represented via an array containing an object of the class \EventInProgress{} for every event in the model. Thereby the distinction between events with and without delays is made. The major difference between both is that an event with delay can trigger multiple times before it is executed. In order to deal with such an issue, the class \SBMLEventInProgressWithDelay{} keeps track of this via the help of a list containing the points in time, at which the respective event has to executed. When events trigger more than once before execution, they have to be ordered according to their delay because the delay of the very same event may vary.

When the \SBMLinterpreter{} is processing events with priority, the events with the highest priority 
are currently stored in list until one of them is selected for execution. One could argue, that all events can
be kept in the same data structure, e.\,g. in a binary max heap, where after the extraction of the element with the largest value, the heap is restructured so that that next largest value is at the top. As stated
in a previous section, the execution of one event can influence the priority off the remaining events. Considering the binary max heap, there is the possibility that many priorities change whereby the standard
method to restore the max heap characteristic after extraction is not sufficient any more. Therefore we disregarded the use of other data structures for now.
%

%The simulation algorithm then proceeds as follows: For each time step, the ODE
%solver gets the current variable values and
%calculates the system's state for the next point in time. After that, events
%and rules are processed, that can change the values. The modified values then
%become the initial values for the next time step. The event processing of the
%Rosenbrock solver
%\sout{is different from other solvers, as it}
%is directly integrated in the solver class and influences the
%step size. The time-accurate handling of events and rules leads to very precise
%results of the simulation.
%
SED-ML support is enabled by inclusion of the \jlibsedml{} library
(\href{http://www.jlibsedml.org}{http://www.jlibsedml.org}) in the binary
download. Clients of the the Simulation Core Library can choose to use the
\jlibsedml{} API directly, or access SED-ML support via  facade classes
in the \texttt{org.simulator.sedml} package that do not require direct
dependencies on \jlibsedml{} in their code.


\section*{Availability and Requirements}
The current version of Simulation Core Library is available on the project homepage. Either the source code or a jar file with binaries can be downloaded.

Project name: Simulation Core Library

Project homepage: http://sourceforge.net/projects/simulation-core/

Operating Systems: Simulation Core Library was successfully tested under Linux, Mac OS X and Windows 7.

Programming Language: Java\texttrademark{}

Other Requirements: Java Runtime Environment 1.6

License: LGPL version 3

\bigskip

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Author's contributions}
RK and AlD contributed equally and declare shared first authorship.

    

%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
The authors are grateful to B.~Kotcon, S.~Mesuro, D.~Rozenfeld, A.~Yodpinyanee,
A.~Perez, E.~Doi, R.~Mehlinger, S.~Ehrlich, M.~Hunt, G.~Tucker, P.~Scherpelz,
A.~Becker, E.~Harley, and C.~Moore, Harvey Mudd College, USA, for providing a
Java implementation of Rosenbrock's method, and to Michael T.~Cooling,
University of Auckland, New Zealand, for fruitful discussion.

This work was funded by the Federal Ministry of Education and Research (BMBF,
Germany) in the project Virtual Liver (grant number 0315756).
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%              
%%  Bmc_article.bst  will be used to                       %%
%%  create a .BBL file for submission, which includes      %%
%%  XML structured for BMC.                                %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %% 
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %% 
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\newpage
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{bmc_article} }     % Bibliography file (usually '*.bib' ) 

%%%%%%%%%%%

\ifthenelse{\boolean{publ}}{\end{multicols}}{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
 \subsection*{Figure 1 - Example for the creation of an abstract syntax graph of a small model}

The model consists of the following reactions:
\begin{eqnarray*}
&R_{1}&: F1,6BP \rightleftharpoons DHAP + GA3P\\
&R_{2}&: DHAP \rightleftharpoons GA3P
\end{eqnarray*}

The two reactions are part of the glycolysis, the contained molecules are Fructose 1,6-bisphosphate (F1,6BP), Dihydroxyacetone phosphate (DHAP) and Glyceraldehyde 3-phosphate (GA3P). With SBMLsqueezer \cite{Draeger2008} the following mass action kinetics have been created:

\begin{eqnarray*}
&v(R_{1})&=k_{ass1} \cdot [F16BP] - k_{diss1} \cdot[DHAP] \cdot [GA3P]\\
&v(R_{2})&=k_{ass2} \cdot [DHAP] - k_{diss2} \cdot[GA3P]
\end{eqnarray*}
The nodes for [DHAP] and [GA3P] are only contained in the syntax graph once and connected to more than one multiplication node. It is obvious from the figure that the syntax graph is not a tree.

 \subsection*{Figure 2 - Algorithm for transforming algebraic rules to assignment rules}
The first step is to decide whether the model is overdetermined by creating a matching of the variables.
If this is not the case, every algebraic rule is solved to the matched variable,
which provides the basis for the creation of an equivalent assignment rule.

 \subsection*{Figure 3 - Processing of events: simpflified algorithm (without delayed events)}
The list of the events running and the inactive events whose trigger is already true are saved.
In each iteration first the trigger of running events that are not persistent are checked.
If the trigger of such an event has changed to false, it is removed from the list.
The next step comprises the evaluation of the triggers of all events.
If its trigger changes from false to true, an event is added to the running events.
An event with its trigger changed from true to false is removed from the list of inactive events.
After the procession of all triggers the event with the highest priority in the list of running events is chosen for execution.
This event is then processed and afterwards the triggers of all events have to be evaluated again.
The algorithm proceeds until the list of running events is empty.

 \subsection*{Figure 4 - Calculation of the derivatives at a specific point in time}
First, the vector for saving the derivatives of all quantities $\dot{\vec{Q}}$ is set to the null vector $\vec{0}$.
Then the rate rules of the model are processed by solving the function $\vec{g}(\vec{Q}, t)$, which can change $\dot{\vec{Q}}$ in some dimensions.
After that for every reaction $R_i$ its velocity $\nu_i$ is computed.
The derivatives of each species (with index $s$) participating in the currently processed reaction $R_i$ are updated
in each step adding the product of the stoichiometry $n_{is}$ and the reaction's velocity $\nu_i$.

 \subsection*{Figure 5 - Integrated calculation of new values for a time step including event processing}
For a certain time interval the Rosenbrock solver always tries to increase time $t$ by the current adaptive
step size $s$ and calculates a new vector of quantities $\vec{Q}_{\mathrm{res}}$.
After a successful step the events and rules of the model are processed.
If theese cause a change in $\vec{Q}$, $h$ is first decreased and then the Rosenbrock solver calculates another vector $\vec{Q}_{t}$ using this adapted step size.
The precision of the event processing is therefore determined by the minimum step size $h_{\min}$.
The $\mathrm{adapt}$ function is defined by Rosenbrock's method \TODO{cite}.

  \subsection*{Figure 6 - Architecture of the Simulation Core Library
  (simplified)}
  
Numerical methods are strictly separated from differential equation systems. The
upper part displays the unified type hierarchy of all currently included numerical integration
methods. The middle part shows the interfaces defining several
special types of the differential equations to be solved by the numerical
methods.
The class \SBMLinterpreter{} (bottom part) implements all of these interfaces
with respect to the information content of a given SBML model. Similarly, an
implementation of further data formats can be included into the
library.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}

  \subsection*{Table 1 a) - Other simulators for SBML models: SBML language support}
    The table shows what SBML elements (fast reactions, algebraic rules, events) are supported by other SBML simulation software and whether all models of the SBML Test Suite are solved correctly. 
	\par \mbox{}
\begin{landscape}
      \begin{tabular}{|l|l|l|l|l|l|l|l|l|l|l|l|l|l|}
\hline
Name 	& Fast reactions & Algebraic rules& Events &Complete SBML Test Suite solved correctly&SED-ML\\\hline
BioUML & Yes& Yes& Yes&Yes&In alpha version\\\hline
COPASI &No & No & Partial &No&No\\\hline
iBioSim &Yes&Yes&Yes&No& No\\\hline
JSim & No & Yes & No &No& No\\\hline
LibSBMLSim &Yes& Yes & Yes&Only for SBML L 3 and SBML L2 V4& No\\\hline
VCell &Yes& No &Yes&No& No\\\hline
Systems Biology Simulation Core Library &Yes& Yes &Yes&Yes& Yes\\\hline
 \end{tabular}
\end{landscape}

\subsection*{Table 1b) - Other simulators for SBML models: special features}
    The table shows general properties of the other simulators for SBML.
	\par \mbox{}
\begin{landscape}
\begin{tabular}{|l|l|l|l|l|l|}
\hline
Name 	&Programming language&API access & Version & Platform & Comments \\\hline
BioUML &Java &JavaScript API&0.9.3 & Platform independent (Java) &\\\hline
COPASI &	C++ with multiple language bindings & Yes &4.8.0& Windows, Linux, Mac OS X, Solaris &\\\hline
iBioSim &Java & No&2.0.0 & Fedora 15, Windows, Mac OS $>$ 10.6 &\\\hline
JSim & Java & Yes &2.07.0 & Windows, Linux, Mac OS X &\\\hline
LibSBMLSim &C & Yes &1.0.0 &Windows, Linux, Free BSD, Mac OS X&\\\hline
VCell &Java frontend, C/C++ server backend & No & 5.0.0 & Platform independent (Java) & Requires internet connection\\\hline
Systems Biology Simulation Core Library &Java & Yes & 1.0 & Platform independent (Java) &\\\hline
 \end{tabular}
 
\end{landscape}

 %Name & URL & Publication& Fast Reactions & Algebraic Rules& Events & Model 966	 & Model 988	& Model 1083 & Model 1000 &Programming language	&API access & License & Native data format & Version & Release & Platform & Comments & SED-ML support\\\hline
%BioUML & http://www.biouml.org/ & \cite{Kolpakov2011}, \cite{Kolpakov2006} & & & & No & &&&&Java & & open source &DML & 0.9.3 & Nov 03 2011 & Platform independent (Java) & Integrators: JVODE, DormandPrince, Radau5, Euler, Imex & Only in alpha version\\\hline
%Cain	1			http://cain.sourceforge.net/									C++ with GUI in wxPython				1.10.0	Sep 27 2011	source code distribution		No
%CompuCell3D	1			http://www.compucell3d.org/									C++ with Python wrapper	Yes		CC3DML	3.6.2		Windows, Linux, Mac OS X > 10.5.8	Could not open SBML file via GUI	No
%COPASI & http://copasi.org/ & \cite{Hoops2006} & No & No & Not every feature supported & Simultaneous events not supported & Fast reactions not supported & Algebraic rules not supported	 & The model uses species reference ids in mathematical expressions. The 5model can currently not be imported by COPASI.&	C++ with multiple language bindings & Yes &Artistic License & CopasiML & 4.8.0	Dec 20 2011	& Windows, Linux, Mac OS X, Solaris & Deterministic integrator: LSODA	& No\\\hline
%GNU MCSim	1			http://www.gnu.org/software/mcsim/									C	Yes	GPL		5.4.0	Jan 29 2011	source code distribution		No
%iBioSim & http://www.async.ece.utah.edu/iBioSim/ & \cite{Myers2009} &	& & &No: could not allocate unit def (null) & Yes &  & Yes & Java & ? & ? & iBioSimModel & 2.0.0 & Aug 26 2011 & Fedora 15, Windows, Mac OS > 10.6 & Simulators: Embedded Runge-Kutta-Fehlberg (4,5) method; Embedded Runge-Kutta Prince-Dormand (8,9) method; Implicit 4th order Runge-Kutta at Gaussian points; Gear method m=2; Gear method m=1; Euler method & No\\\hline
%insilicoIDE	1			http://www.physiome.jp/									C++, Qt			insilicoML	1.4.4	Nov 18 2011	Windows, Linux, FreeBSD, MacOS > 10.5	Import of SBML models hidden under menu entry "Item", no import possible because no reaction upon clicking on import SBML. Two integration methods available: Euler and Runge-Kutta	No
%JSim & http://www.physiome.org/jsim/ &  \cite{Beard2012a} & Yes & & &	& & &	Wrong & Java & & open source & MML (Mathematical Modeling Language) & 2.07.0 & Jun 12 2012 & Windows, Linux, Mac OS X && No\\\hline
%JWS Online & http://jjj.biochem.sun.ac.za/index.html & \cite{Olivier2004} & & & & & & & & Java & & & JWS Online Format (*.dat) & & Online program & Supports only SBML Level 1 and Level 2. Models must be uploaded and can only be simulated online; no stand-alone version & No \\\hline
%MOOSE	1			http://moose.ncbs.res.in/									Python		LGPL			Jan 20 2012	Linux, Windows		
%NetBuilder	1			http://strc.herts.ac.uk/bio/maria/Apostrophe/									Python	Yes	LGPL	SBML	0.5.0	Sep 03 2011	Windows		
%PySCeS	1			http://pysces.sourceforge.net/	Olivier2005								Python		GPL				Windows, Linux	Requires Fortran	
%SOSlib	1			http://www.tbi.univie.ac.at/~raim/odeSolver/	Machne2006								C	Yes	LGPL	SBML	1.6.0	Dec 17 2005	source code distribution		No
%SYCAMORE	1			http://sycamore.eml.org/	Weidemann2008											SBML			Online program	Internally uses COPASI for simulation	No
%LibSBMLSim &http://fun.bio.keio.ac.jp/software/libsbmlsim/& \cite{Moraru2008}, \cite{ Loew2001} & Yes & & &Error adding features & Cannot generate math & & No rate rules supported & Java frontend C/C++ server backend & No & open source & VCML	& 5.0.0 & Nov 11 2011 & Platform independent (Java) & Requires internet connection & 

%VCell & http://www.vcell.org/ & \cite{Moraru2008}, \cite{ Loew2001} & Yes & & &Error adding features & Cannot generate math & & No rate rules supported & Java frontend C/C++ server backend & No & open source & VCML	& 5.0.0 & Nov 11 2011 & Platform independent (Java) & Requires internet connection & No\\\hline
%XPPAUT	1			http://www.math.pitt.edu/~bard/xpp/xpp.html	Ermentrout2012		
\subsection*{Table 2 - Simulation of the models from the SBML Test Suite (version 2.0.2) using Rosenbrock solver}
The table shows for all SBML levels and versions the number of tested models and the total running times of the tests.
An Intel\textsuperscript{\textregistered} Core\texttrademark{} i5 CPU with 3.33\,GHz  and 4\,GB RAM was used with
Windows\textsuperscript{\textregistered}~7 (Version 6.1.7600) as operating system.\par \mbox{}
\par
    \mbox{
\rowcolors{2}{white}{lightblue}
\begin{tabular}{rrrrr}
\toprule
\multicolumn{1}{c}{Level} & \multicolumn{1}{c}{Version} &
\multicolumn{1}{c}{Models} & \multicolumn{1}{c}{Correct simulations} &
\multicolumn{1}{c}{Total running time (in s)}\\
\midrule
1 & 2 &   252 &   252 &  1.1\\
2 & 1 &   885 &   885 &  3.6\\
2 & 2 & 1,039 & 1,039 &  3.3\\
2 & 3 & 1,039 & 1,039 &  3.2\\
2 & 4 & 1,041 & 1,041 &  3.2\\
3 & 1 & 1,075 & 1,075 & 34.3\\
\bottomrule
\end{tabular}
}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- User Guide}
    The user guide shows how to read in a model and how to conduct a simulation afterwards.

\end{bmcformat}
\end{document}
